### PR TITLE
fix cache invalidations when result contains empty values

### LINKED ISSUE #1953: @envelop/response-cache clarify handling of null or empty values
Recently we've seen some issues with stale data coming from our GraphQL server. It's pretty green field so the general entities involved could be anything, so let's just say we have `Product`, `Query.product` and `Mutation.createProduct`:

```graphql
type Query {
  products: [Product!]!
}

type Mutation {
  addProduct(input: ProductInput!): Product!
}

input ProductInput {
  name: String!
}

type Product {
  id: ID!
  name: String!
}
```

Whenever we handle `createProduct`, we invoke `cache.invalidate([typename: "Product"])`, so every operation related to `Product` would get invalidated - or so we thought. Turns out this only invalidates operations where a `Product` entity has been included in the response. If `Query.products` at some point returned an empty list, the response would not get invalidated on a subsequent `Mutation.createProduct` invocation. An exemplary test in `response-cache.spec.ts` should clarify:

```js
  it('purges cached empty query operation execution result via imperative cache.invalidate api using typename', async () => {
    const spy = jest.fn(() => []);
    let id = 1;
    const schema = makeExecutableSchema({
      typeDefs: /* GraphQL */ `
        type Query {
          products: [Product!]!
        }

        type Mutation {
          addProduct(input: ProductInput!): Product!
        }

        input ProductInput {
          name: String!
        }

        type Product {
          id: ID!
          name: String!
        }
      `,
      resolvers: {
        Query: {
          products: spy,
        },
        Mutation: {
          addProduct(_, { input }) {
            const mockProduct = {
              id: id++,
              name: input.name,
            };
            return mockProduct;
          },
        },
      },
    });

    const cache = createInMemoryCache();
    const testInstance = createTestkit([useResponseCache({ session: () => null, cache })], schema);

    const query = /* GraphQL */ `
      query test {
        products {
          id
          name
        }
      }
    `;

    await testInstance.execute(query);
    await testInstance.execute(query);
    expect(spy).toHaveBeenCalledTimes(1);

    cache.invalidate([{ typename: 'Product' }]);

    await testInstance.execute(query);
    expect(spy).toHaveBeenCalledTimes(2);
  });
```
```
  ● useResponseCache › purges cached empty query operation execution result via imperative cache.invalidate api using typename

    expect(jest.fn()).toHaveBeenCalledTimes(expected)

    Expected number of calls: 2
    Received number of calls: 1

      532 |
      533 |     await testInstance.execute(query);
    > 534 |     expect(spy).toHaveBeenCalledTimes(2);
          |                 ^
      535 |   });
      536 |
      537 |   it('variables are used for constructing the cache key', async () => {

      at Object.toHaveBeenCalledTimes (packages/plugins/response-cache/test/response-cache.spec.ts:534:17)
```

Is there a recommended approach to this? I'm tempted to give a fix a try, but it'd rather wait for feedback because it might as well be intended design. Technically, I think we could parse the schema, track which fields have what type and use that to invalidate results with null / empty values.

### PR DESCRIPTION
**Pull request by marhaupe**
🚨 **IMPORTANT: Please do not create a Pull Request without creating an issue first.**

_Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of
the pull request._

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Fixes #1953 (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Issue #1953 contains most relevant information, also I added a new test `purges cached empty query operation execution result via imperative cache.invalidate api using typename`.

**Test Environment**:

- OS: MacOS
- `@envelop/...`:
- NodeJS: 18.16.0

## Checklist:

- [x] I have followed the
      [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the
      style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

This might as well just be a draft. I'm not sure if I need to follow the `CONTRIBUTING.md` for drafts. Let me know if this is useful or not.

### CODE REVIEW

File: packages/plugins/response-cache/src/plugin.ts [hunk 1]
```diff
@@ -295,11 +296,10 @@ export function useResponseCache<PluginContext extends Record<string, any> = {}>
       if (processedSchemas.has(schema)) {
         return;
       }
-      // Check if the schema has @cacheControl directive
       const cacheControlDirective = schema.getDirective('cacheControl');
-      if (cacheControlDirective) {
```
Comment (ardatan) [inline packages/plugins/response-cache/src/plugin.ts:300]: Is this really needed because this visits over the schema even if `cacheControl` is not present.

Comment (marhaupe) [reply packages/plugins/response-cache/src/plugin.ts:300]: Yeah this is by no means ready to merge, I honestly did not want to risk optimizing a solution that I'm not sure is actually wanted. If we're already talking about it though, what would you suggest? I would probably try to iterate over the schema twice to keep things clean. Keep the handling for `@cacheControl` as it is, and generate the `Map` in a separate iteration separately, depending on `invalidateViaMutations`

**Comment by EmrysMyrddin** [comment]
Thank you for sharing this with us !

Honestly, I'm pretty happy with your proposal. I think that it is a good improvement, the current behaviour of the cache invalidation for empty values is actually a bug we need to fix I think.

Your approach seems totally fine for me !

I will add some observations but I think this is totally something we want to merge :-)

File: packages/plugins/response-cache/src/plugin.ts [hunk 2]
```diff
@@ -310,9 +310,14 @@ export function useResponseCache<PluginContext extends Record<string, any> = {}>
...
               }
             });
-            return type;
-          },
-          [MapperKind.FIELD]: (fieldConfig, fieldName, typeName) => {
+          }
+          return type;
+        },
+        [MapperKind.FIELD]: (fieldConfig, fieldName, typeName) => {
+          const id = `${typeName}.${fieldName}`;
```
Comment (EmrysMyrddin) [inline packages/plugins/response-cache/src/plugin.ts]: Perhaps call this variable `schemaCoordinates` ?
```suggestion
          const schemaCoordinates = `${typeName}.${fieldName}`;
```

File: packages/plugins/response-cache/test/response-cache.spec.ts [hunk 3]
```diff
@@ -480,6 +480,66 @@ describe('useResponseCache', () => {
     expect(spy).toHaveBeenCalledTimes(2);
   });
 
+  it('purges cached empty query operation execution result via imperative cache.invalidate api using typename', async () => {
```
Comment (EmrysMyrddin) [inline packages/plugins/response-cache/test/response-cache.spec.ts]: Can you add a test for the null case too ?

File: packages/plugins/response-cache/src/plugin.ts (hunk 1 above)
Comment (EmrysMyrddin) [reply packages/plugins/response-cache/src/plugin.ts:300]: I think that it's ok like this. If you want to make things more clear, you can eventually create functions with meaningful names to make it clear what is done in each visitor hook. But visiting the entire schema twice can hurt initialisation performance and hurt the startup time :-)

Comment (EmrysMyrddin) [reply packages/plugins/response-cache/src/plugin.ts:300]: And I think we should always build the Map, even when automatic cache invalidation is turned off, since we still want to be able to manually invalidate the cache.

**Review by EmrysMyrddin — CHANGES_REQUESTED** [review]

**Comment by marhaupe** [comment]
Thanks for your quick feedback! I've just pushed some changes, pretty happy with how it's turned out.

File: packages/plugins/response-cache/src/plugin.ts (hunk 1 above)
Comment (marhaupe) [reply packages/plugins/response-cache/src/plugin.ts:300]: Right now I'm not visiting `MapperKind.COMPOSITE_TYPE` if `cacheControlDirective` does not exist. I think this is as good as it gets for now.

**Comment by EmrysMyrddin** [comment]
Thank you for taking the time for this 😀 

After some thoughts, I see a drawback with our approach here.

The idea here is to be able to invalidate by type when we add an entity to avoid having null or lists stale. But this comes at cost of loosing the cache for any response which contains at least one entity of this type, while it is probably not stale just because we added an item.

It is still fine to merge this, I should probably open a new issue to discuss about this problem, it's probably more a lack of api for the invalidation part of the cache.

**Review by EmrysMyrddin — APPROVED** [review]

**Comment by EmrysMyrddin** [comment]
I've created an issue to discuss about my concern here : #1979 

Let's merge this since it's not really related.

@marhaupe Can you add a changeset ? Just run `pnpm changeset` and follow instructions :-)

**Comment by marhaupe** [comment]
Done 👍 I also read through your proposal and I like it, I just left a comment with one thought I had though.

**Comment by EmrysMyrddin** [comment]
Thank you again for the time spent on this :-) very appreciated !
